### PR TITLE
Phase 3 R6: Curriculum Learning — Easy-to-Hard Training Strategy (8 parallel)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis — PLATEAU PROTOCOL
After R5 showed 0 improvements across 63 experiments, we activate the Plateau Protocol. Curriculum learning trains on easy samples first, then gradually introduces harder ones. This could help the model build better representations before tackling difficult OOD/tandem cases.

**Baseline flags:**
```bash
--field_decoder --adaln_output --use_lion --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --lr 2e-4 --n_layers 3 --slice_num 96
```

## Instructions
Pull latest noam. SENPAI_TIMEOUT_MINUTES=180, SENPAI_MAX_EPOCHS=500. Use `--wandb_group "phase3-r6-curriculum"`.

### Curriculum Strategy
The idea: sort training samples by difficulty (e.g., by surface pressure variance, or by tandem vs single-foil) and train on easy samples first. Implementation approaches:

### GPU 0: Single-foil only for first 50 epochs, then full dataset
Skip tandem samples for the first 50 epochs. This gives the model a clean learning signal on simpler geometries.
```bash
CUDA_VISIBLE_DEVICES=0 python train.py --field_decoder --adaln_output --use_lion --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --lr 2e-4 --n_layers 3 --slice_num 96 --wandb_name "frieren/r6-curriculum-single-first" --wandb_group "phase3-r6-curriculum" --agent frieren
```

### GPU 1: Volume-only loss for first 30 epochs, then add surface loss
Train on volume nodes first to learn the global flow field, then add surface loss.
```bash
CUDA_VISIBLE_DEVICES=1 python train.py --field_decoder --adaln_output --use_lion --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --lr 2e-4 --n_layers 3 --slice_num 96 --wandb_name "frieren/r6-curriculum-vol-first" --wandb_group "phase3-r6-curriculum" --agent frieren
```

### GPU 2: Progressive surface weight (0.1→10.0 over training)
Start with low surface weight and gradually increase. Opposite of current approach.
```bash
CUDA_VISIBLE_DEVICES=2 python train.py --field_decoder --adaln_output --use_lion --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --lr 2e-4 --n_layers 3 --slice_num 96 --wandb_name "frieren/r6-progressive-surf" --wandb_group "phase3-r6-curriculum" --agent frieren
```

### GPU 3: Sort samples by Re number (low Re first, high Re later)
Low-Re flows are simpler (more laminar). Introduce high-Re flows after epoch 40.
```bash
CUDA_VISIBLE_DEVICES=3 python train.py --field_decoder --adaln_output --use_lion --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --lr 2e-4 --n_layers 3 --slice_num 96 --wandb_name "frieren/r6-curriculum-re" --wandb_group "phase3-r6-curriculum" --agent frieren
```

### GPU 4: Anti-curriculum (hard samples first)
Train on tandem samples only for first 50 epochs. If the model can learn the hardest cases first, single-foil may come easier.
```bash
CUDA_VISIBLE_DEVICES=4 python train.py --field_decoder --adaln_output --use_lion --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --lr 2e-4 --n_layers 3 --slice_num 96 --wandb_name "frieren/r6-anti-curriculum" --wandb_group "phase3-r6-curriculum" --agent frieren
```

### GPU 5: Error-weighted sampling (oversample high-error samples)
After epoch 40, compute per-sample errors and oversample the worst-performing samples 3x.
```bash
CUDA_VISIBLE_DEVICES=5 python train.py --field_decoder --adaln_output --use_lion --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --lr 2e-4 --n_layers 3 --slice_num 96 --wandb_name "frieren/r6-error-sampling" --wandb_group "phase3-r6-curriculum" --agent frieren
```

### GPU 6: Two-phase LR (lr=5e-4 for first 100 epochs, then lr=1e-4)
Aggressive learning first, then refinement. Different from cosine schedule.
```bash
CUDA_VISIBLE_DEVICES=6 python train.py --field_decoder --adaln_output --use_lion --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --lr 2e-4 --n_layers 3 --slice_num 96 --wandb_name "frieren/r6-two-phase-lr" --wandb_group "phase3-r6-curriculum" --agent frieren
```

### GPU 7: Progressive noise reduction (start with 2x noise, reduce to 0)
Heavy regularization early, then clean signal for final convergence.
```bash
CUDA_VISIBLE_DEVICES=7 python train.py --field_decoder --adaln_output --use_lion --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --lr 2e-4 --n_layers 3 --slice_num 96 --wandb_name "frieren/r6-progressive-noise" --wandb_group "phase3-r6-curriculum" --agent frieren
```

## Baseline
| val/loss | p_in | p_oodc | p_tan | p_re |
|----------|------|--------|-------|------|
| **0.3997** | 12.7 (lucky) / ~13.8 (mean) | 8.8 | 33.2 | 24.8 |

---

### Setup
- 8 parallel GPU runs, 3h each, baseline flags: `--field_decoder --adaln_output --use_lion --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --lr 2e-4 --n_layers 3 --slice_num 96`

### Results Table

| GPU | Strategy | W&B Run | val/loss | p_in | p_oodc | p_tan | p_re | Peak Mem |
|-----|----------|---------|----------|------|--------|-------|------|----------|
| 0 | curriculum_single_first | 4g9fwtng | 0.4145 | 13.9 | 8.6 | 34.8 | 24.6 | 34.7 GB |
| 1 | curriculum_vol_first | yecafj86 | 0.4090 | 13.6 | 8.7 | 34.3 | 24.5 | 38.3 GB |
| 2 | curriculum_progressive_surf | j55f7wk2 | **0.4022** | 13.8 | 8.6 | **33.2** | **24.3** | 37.2 GB |
| 3 | curriculum_re | 4qmvbql5 | 0.5989 | 18.4 | 15.1 | 48.8 | 27.5 | 34.3 GB |
| 4 | curriculum_tandem_first | il44e0lz | 0.4062 | 13.8 | 8.3 | 33.9 | 24.7 | 36.8 GB |
| 5 | curriculum_error_sampling | ssay8e7w | 0.4031 | 13.5 | 8.5 | 32.9 | 24.6 | 37.1 GB |
| 6 | two_phase_lr | 0wwmzceb | 0.6096 | 18.2 | 15.9 | 48.0 | 28.6 | 36.3 GB |
| 7 | progressive_noise | kucu3pjb | 0.4108 | 14.0 | 8.1 | 34.2 | 24.3 | 38.3 GB |
| **Baseline** | 3L+96s+high_p_clamp | hj8x5227 | **0.3997** | 13.8 | 8.8 | 33.2 | 24.8 | — |

All values are best checkpoint metrics. p_in/oodc/tan/re = mae_surf_p (Pa) on respective val splits.

### What Happened

**None of the curriculum strategies beat the baseline (0.3997)** in the 3h budget. Results cluster into two groups:

**Competitive (within 3% of baseline):**
- GPU2 (progressive_surf, 0.4022): Best overall. Linear surf_weight ramp 0.1 to 10.0 over 200 epochs is the closest to matching baseline. p_tan=33.2 (tied with baseline). Starting with low surface emphasis helps volume learning first without hurting final accuracy.
- GPU5 (error_sampling, 0.4031): Best p_tan at 32.9, marginally better than baseline (33.2) on the hardest split. Error-weighted sampling after epoch 40 provides a small but consistent signal.
- GPU4 (tandem_first, 0.4062): Anti-curriculum surprisingly competitive. Training tandem-heavy for 50 epochs then switching to full dataset seems fine but doesn't help.
- GPU1 (vol_first, 0.4090): Setting surf_weight=0 for 30 epochs then adaptive is close but not better.
- GPU7 (progressive_noise, 0.4108): 2x noise reducing to 0 over 150 epochs makes marginal difference.
- GPU0 (single_first, 0.4145): Skipping tandem batches for 50 epochs hurts slightly. Model learns single-foil geometry well but transfers less effectively to tandem by epoch 200.

**Clearly worse:**
- GPU3 (curriculum_re, 0.5989): Withholding high-Re training data for 40 epochs causes persistent accuracy degradation across all splits. Re is a key conditioning variable and withholding it prevents the model from learning the full distribution.
- GPU6 (two_phase_lr, 0.6096): High initial lr=5e-4 (2.5x above our optimal 2e-4) causes poor convergence. The step transition at epoch 100 is too coarse. Our baseline Lion + cosine annealing schedule is clearly superior.

**Implementation note:** The existing `tandem_curriculum_epochs=10` feature had a pre-existing bug. After Fourier PE was appended to x, the foil-2 detection check `x[:, :, -8:]` inadvertently checks Fourier cosine features instead of foil-2 DSDF channels, treating all samples as "tandem" and zeroing all gradients for epochs 0-9. This affected all 8 runs equally (first 10 epochs are effectively no-ops), so relative comparison is valid, but the effective training window was ~160 epochs instead of the full 180.

### Suggested Follow-Ups

1. **Progressive surface weight** (GPU2) is the most promising — it self-organizes loss weighting and avoids the sensitivity of the adaptive ratio approach. Worth trying longer ramp (0.1 to 20.0 over 300 epochs) or combining with error sampling.
2. **Error sampling** (GPU5) shows the best p_tan (32.9) — slight improvement on the hardest split. Could be combined with progressive_surf or run with a longer budget.
3. **Fix the tandem_curriculum bug**: The `is_tandem_curr` check at line ~1243 uses `x[:, :, -8:]` which now catches Fourier PE features. Fix: use the standardized gap feature (`x[:, 0, 21].abs() > threshold`) to correctly identify tandem samples. This would recover 10 epochs of effective training.
4. **Re-based filtering and two-phase LR** should not be pursued further — both clearly hurt.

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT: I've read the PR carefully and noticed all 8 GPU commands are identical (same flags, just different --wandb_name). None of the curriculum strategies (single-foil first, volume-only loss, progressive surface weight, Re sorting, anti-curriculum, error-weighted sampling, two-phase LR, progressive noise) have corresponding CLI flags in the current train.py.

Could you clarify your intent? Specifically:
1. Should I implement new CLI flags for each curriculum strategy (e.g., `--curriculum_tandem_start 50`, `--surf_start_epoch 30`, `--progressive_surf_weight`, etc.)?
2. Or did you intend to provide specific flags that are already supported but forgot to include them in the commands?

GPU 6 (two-phase LR) is the easiest as it could use a custom scheduler, and GPU 2 (progressive surface weight) could use a ramp similar to existing `tandem_ramp`. But GPUs 0 (single-foil first), 3 (Re sorting), 4 (anti-curriculum), and 5 (error-weighted sampling) would require new dataset-level or sampler-level code.

Please let me know if I should proceed with implementing new features, and if so, which are the highest priority.

**Comment by tcapelle** [comment]
ADVISOR: Good question! Yes — these curriculum strategies require CODE CHANGES to train.py. You need to implement the logic for each variant. Here's how:

For each GPU, you'll add the curriculum logic directly in the training loop. The key modifications are:

### GPU 0 (single-foil first): 
In the training loop, add a check: `if epoch < 50 and is_tandem.any(): continue`. Skip batches containing tandem samples during early epochs. Add a `--curriculum_single_first` flag.

### GPU 1 (volume-only first):
For `epoch < 30`, set `surf_weight = 0.0`. Override the adaptive surf_weight with zero for early epochs. Add `--curriculum_vol_first` flag.

### GPU 2 (progressive surface weight):
Replace the adaptive surf_weight with a linear ramp: `surf_weight = 0.1 + (10.0 - 0.1) * min(1.0, epoch / 200)`. Add `--curriculum_progressive_surf` flag.

### GPU 5 (error-weighted sampling):
After epoch 40, compute per-sample validation errors, create a WeightedRandomSampler that oversamples high-error samples. Add `--curriculum_error_sampling` flag.

### GPU 6 (two-phase LR):
Set lr=5e-4 for epochs 0-100, then lr=1e-4 for 100+. Modify the scheduler setup. Add `--two_phase_lr` flag.

### GPU 7 (progressive noise):
Modify `noise_scale = 0.10 * max(0, 1 - epoch / 150)`. Start with 2x noise, linearly reduce to 0. Add `--progressive_noise` flag.

For the simpler variants (GPU 3, 4), you can implement them by filtering the dataset based on metadata.

Each variant is a small, focused code change. Implement each behind its own flag so they can be independently tested. Proceed with implementation and run all 8 in parallel.
